### PR TITLE
Modify Calculate Fee

### LINKED ISSUE #2413: if you are a CalculateAlgorithmService, why do you need to care about fork? one ...
https://github.com/AElfProject/AElf/blob/40ce8b375b50421ee20b59c003701f91098fa679/src/AElf.Kernel.SmartContract/Application/CalculateFeeStrategy/ICalculateAlgorithmService.cs#L10-L13

---

###### This issue was generated by [todo](https://todo.jasonet.co) based on a `TODO` comment in 40ce8b375b50421ee20b59c003701f91098fa679 when #2409 was merged. cc @AElfProject.

### LINKED ISSUE #2414: we only need one strategy to calculate tx, read, write ... resources.
https://github.com/AElfProject/AElf/blob/40ce8b375b50421ee20b59c003701f91098fa679/src/AElf.Kernel.SmartContract/Application/CalculateFeeStrategy/ICalculateCostStrategy.cs#L15-L18

---

###### This issue was generated by [todo](https://todo.jasonet.co) based on a `TODO` comment in 40ce8b375b50421ee20b59c003701f91098fa679 when #2409 was merged. cc @AElfProject.

### LINKED ISSUE #2415: it looks not OOP
https://github.com/AElfProject/AElf/blob/40ce8b375b50421ee20b59c003701f91098fa679/src/AElf.Kernel.SmartContract/Application/CalculateFeeStrategy/ICalculateWay.cs#L9-L12

---

###### This issue was generated by [todo](https://todo.jasonet.co) based on a `TODO` comment in 40ce8b375b50421ee20b59c003701f91098fa679 when #2409 was merged. cc @AElfProject.

### LINKED ISSUE #2418: why in a SmartContract package, you put a Token class? if you create a door, can...
https://github.com/AElfProject/AElf/blob/40ce8b375b50421ee20b59c003701f91098fa679/src/AElf.Kernel.SmartContract/Application/ExtraAvailableTokenProvider.cs#L3-L7

---

###### This issue was generated by [todo](https://todo.jasonet.co) based on a `TODO` comment in 40ce8b375b50421ee20b59c003701f91098fa679 when #2409 was merged. cc @AElfProject.

### LINKED ISSUE #2419: if you new a class, why not pass parameters by constructor? you know the type of...
https://github.com/AElfProject/AElf/blob/40ce8b375b50421ee20b59c003701f91098fa679/src/AElf.Kernel.TransactionPool/Application/CalculateFeeService.cs#L165-L168

---

###### This issue was generated by [todo](https://todo.jasonet.co) based on a `TODO` comment in 40ce8b375b50421ee20b59c003701f91098fa679 when #2409 was merged. cc @AElfProject.

### LINKED ISSUE #2442: service should not expose a Context as a property
https://github.com/AElfProject/AElf/blob/a5bc2e485ed926d696ea4a3014f4050c492434aa/src/AElf.Kernel.SmartContract/Application/CalculateFeeStrategy/ICalculateAlgorithmService.cs#L7-L10

---

###### This issue was generated by [todo](https://todo.jasonet.co) based on a `TODO` comment in a5bc2e485ed926d696ea4a3014f4050c492434aa when #2426 was merged. cc @AElfProject.

### LINKED ISSUE #2445: smart contract should not know fee, fee is just a plugin
https://github.com/AElfProject/AElf/blob/a5bc2e485ed926d696ea4a3014f4050c492434aa/src/AElf.Kernel.SmartContract/Application/ISymbolListToPayTxFeeService.cs#L6-L11

---

###### This issue was generated by [todo](https://todo.jasonet.co) based on a `TODO` comment in a5bc2e485ed926d696ea4a3014f4050c492434aa when #2426 was merged. cc @AElfProject.

### LINKED ISSUE #2453: if only add something to cache, the method name should call AddAlgorithmCacheByB...
https://github.com/AElfProject/AElf/blob/a5bc2e485ed926d696ea4a3014f4050c492434aa/src/AElf.Kernel.TransactionPool/Application/CalculateFeeService.cs#L87-L90

---

###### This issue was generated by [todo](https://todo.jasonet.co) based on a `TODO` comment in a5bc2e485ed926d696ea4a3014f4050c492434aa when #2426 was merged. cc @AElfProject.

### LINKED ISSUE #2460: change strategy implement
https://github.com/AElfProject/AElf/blob/c63d5aadc6cc9148eb7cc41f7c18d71cd7f1c39d/src/AElf.Kernel.SmartContract.ExecutionPluginForAcs8/ResourceConsumptionPostExecutionPlugin.cs#L26-L29

---

###### This issue was generated by [todo](https://todo.jasonet.co) based on a `TODO` comment in c63d5aadc6cc9148eb7cc41f7c18d71cd7f1c39d when #2459 was merged. cc @AElfProject.

### LINKED ISSUE #2480: should not implement here
https://github.com/AElfProject/AElf/blob/c63d5aadc6cc9148eb7cc41f7c18d71cd7f1c39d/src/AElf.Kernel.TransactionPool/Application/CalculateCostStrategy.cs#L19-L22

---

###### This issue was generated by [todo](https://todo.jasonet.co) based on a `TODO` comment in c63d5aadc6cc9148eb7cc41f7c18d71cd7f1c39d when #2459 was merged. cc @AElfProject.

### LINKED ISSUE #2481: should not implement here, no fork
https://github.com/AElfProject/AElf/blob/c63d5aadc6cc9148eb7cc41f7c18d71cd7f1c39d/src/AElf.Kernel.TransactionPool/Application/DeployContractAddressForkCacheHandler.cs#L8-L12

---

###### This issue was generated by [todo](https://todo.jasonet.co) based on a `TODO` comment in c63d5aadc6cc9148eb7cc41f7c18d71cd7f1c39d when #2459 was merged. cc @AElfProject.

### LINKED ISSUE #2484: not here
https://github.com/AElfProject/AElf/blob/c63d5aadc6cc9148eb7cc41f7c18d71cd7f1c39d/src/AElf.Kernel.TransactionPool/Application/SymbolListToPayTxFeeForkCacheHandler.cs#L8-L11

---

###### This issue was generated by [todo](https://todo.jasonet.co) based on a `TODO` comment in c63d5aadc6cc9148eb7cc41f7c18d71cd7f1c39d when #2459 was merged. cc @AElfProject.

### PR DESCRIPTION
**Pull request by llypb**
- [x]  re-post calculate fee code fix #2445  fix #2480

- [x] remove function provider, ICalculateWay fix #2413  fix #2415  fix #2419 fix #2442 fix #2453

- [x] add IResourceTokenFeeProvider instead of ICalculateFeeStradegy fix #2414 fix #2460 fix #2484 

- [x] delete the token class after modifying fork cache fix #2418

- [x] change data structure  about coefficient 

fix #2481

### CODE REVIEW

File: protobuf/token_contract.proto [hunk 1]
```diff
@@ -446,7 +443,8 @@ message TransactionSizeFeeUnitPriceUpdated {
 
 message NoticeUpdateCalculateFeeAlgorithm {
     option (aelf.is_event) = true;
-    CalculateFeeCoefficientsOfType all_coefficient = 1;
+    sint32 fee_type = 1;
+    CalculateFeeCoefficientsOfType all_coefficient = 2;
```
Comment (EanCuznaivy) [inline protobuf/token_contract.proto]: all_coefficients

File: src/AElf.Kernel.FeeCalculation/TokenFeeProvider/Impl/TrafficFeeProvider.cs [hunk 2]
```diff
@@ -0,0 +1,27 @@
+using AElf.Contracts.MultiToken;
+using AElf.Kernel.FeeCalculation.ResourceTokenFeeProvider.Impl;
+using AElf.Kernel.SmartContract.Sdk;
+
+namespace AElf.Kernel.FeeCalculation.Impl
```
Comment (EanCuznaivy) [inline src/AElf.Kernel.FeeCalculation/TokenFeeProvider/Impl/TrafficFeeProvider.cs]: use full name Implementation

File: src/AElf.Kernel.TransactionPool/AElf.Kernel.TransactionPool.csproj [hunk 3]
```diff
@@ -10,6 +10,7 @@
     </PropertyGroup>
 
     <ItemGroup>
+        <ProjectReference Include="..\AElf.Kernel.FeeCalculation\AElf.Kernel.FeeCalculation.csproj" />
```
Comment (loning) [inline src/AElf.Kernel.TransactionPool/AElf.Kernel.TransactionPool.csproj]: dont reference any new project

File: src/AElf.Kernel.TransactionPool/Application/TransactionFeeCalculatorCoefficientUpdatedLogEventProcessor.cs [hunk 4]
```diff
@@ -1,10 +1,7 @@
-using System;
-using System.Collections.Generic;
-using System.Linq;
-using System.Threading.Tasks;
+﻿using System.Threading.Tasks;
```
Comment (EanCuznaivy) [inline src/AElf.Kernel.TransactionPool/Application/TransactionFeeCalculatorCoefficientUpdatedLogEventProcessor.cs]: Move calculation files to FeeCalculation project.

File: src/AElf.Kernel.TransactionPool/TransactionPoolAElfModule.cs [hunk 5]
```diff
@@ -8,7 +8,8 @@
 
 namespace AElf.Kernel.TransactionPool
 {
-    [DependsOn(typeof(CoreKernelAElfModule))]
+    [DependsOn(typeof(CoreKernelAElfModule),
+        typeof(FeeCalculationModule))]
```
Comment (EanCuznaivy) [inline src/AElf.Kernel.TransactionPool/TransactionPoolAElfModule.cs]: Remove this ref.

File: src/AElf.Kernel/NewIrreversibleBlockFoundEventHandler.cs [hunk 6]
```diff
@@ -29,7 +31,8 @@ public class NewIrreversibleBlockFoundEventHandler : ILocalEventHandler<NewIrrev
             ITransactionBlockIndexService transactionBlockIndexService, 
             IForkCacheService forkCacheService,
             IChainBlockLinkService chainBlockLinkService,
-            ISmartContractExecutiveService smartContractExecutiveService)
+            ISmartContractExecutiveService smartContractExecutiveService,
+            ICoefficientsCacheProvider coefficientsCacheProvider)
```
Comment (EanCuznaivy) [inline src/AElf.Kernel/NewIrreversibleBlockFoundEventHandler.cs]: Need a solution to manage cache.

File: src/AElf.Kernel.FeeCalculation/Application/ICalculateFunctionProvider.cs [hunk 7]
```diff
@@ -0,0 +1,29 @@
+﻿using System;
```
Comment (EanCuznaivy) [inline src/AElf.Kernel.FeeCalculation/Application/ICalculateFunctionProvider.cs]: Reorganize this project, and add Application to namespances.

File: src/AElf.Kernel.TransactionPool/Application/SymbolListToPayTxFeeUpdatedLogEventProcessor.cs [hunk 8]
```diff
@@ -11,8 +11,6 @@
 
 namespace AElf.Kernel.TransactionPool.Application
 {
-    //TODO: not here
```
Comment (rosona) [inline src/AElf.Kernel.TransactionPool/Application/SymbolListToPayTxFeeUpdatedLogEventProcessor.cs]: only remove the todo, not change the code.

File: src/AElf.Kernel.FeeCalculation/Infrastructure/TokenFeeProviderBase.cs [hunk 9]
```diff
@@ -0,0 +1,62 @@
...
+        }
+
+        public void UpdatePieceWiseFunction(int[] pieceTypeArray)
+        {
+            PieceCalculateFunction = new PieceCalculateFunction();
+            foreach (var pieceType in pieceTypeArray)
+            {
+                switch (pieceType)
+                {
+                    case 0 :
```
Comment (loning) [inline src/AElf.Kernel.FeeCalculation/Infrastructure/TokenFeeProviderBase.cs]: make a Factory pattern to get function, not case

File: test/AElf.Kernel.SmartContract.ExecutionPluginForCallThreshold.Tests/MockCoefficientProvider.cs [hunk 10]
```diff
@@ -0,0 +1,67 @@
...
+using Volo.Abp.DependencyInjection;
+
+namespace AElf.Kernel.SmartContract.ExecutionPluginForCallThreshold.Tests
+{
+    public class MockCoefficientProvider : ICoefficientsCacheProvider, ISingletonDependency
+        {
+            private readonly Dictionary<int, IList<int[]>> _coefficientsDicCache;
+            public MockCoefficientProvider()
+            {
+                _coefficientsDicCache = new Dictionary<int, IList<int[]>>();
```
Comment (loning) [inline test/AElf.Kernel.SmartContract.ExecutionPluginForCallThreshold.Tests/MockCoefficientProvider.cs]: No hard code in class

File: test/AElf.Kernel.SmartContract.ExecutionPluginForResourceFee.Tests/MockCeofficientProvider.cs [hunk 11]
```diff
@@ -0,0 +1,68 @@
...
+namespace AElf.Kernel.SmartContract.ExecutionPluginForResourceFee.Tests
+{
+    public class MockCoefficientProvider : ICoefficientsCacheProvider, ISingletonDependency
+    {
+        private readonly Dictionary<int, IList<int[]>> _coefficientsDicCache;
+
+        public MockCoefficientProvider()
+        {
+            _coefficientsDicCache = new Dictionary<int, IList<int[]>>();
+            var txCoefficient = new List<int[]>
```
Comment (loning) [inline test/AElf.Kernel.SmartContract.ExecutionPluginForResourceFee.Tests/MockCeofficientProvider.cs]: Don't copy any codes!

File: test/AElf.Kernel.SmartContract.ExecutionPluginForMethodFee.Tests/MockCoefficientProvider.cs [hunk 12]
```diff
@@ -0,0 +1,68 @@
...
+
+namespace AElf.Kernel.SmartContract.ExecutionPluginForMethodFee.Tests
+{
+    public class MockCoefficientProvider : ICoefficientsCacheProvider, ISingletonDependency
+    {
+        private readonly Dictionary<int, IList<int[]>> _coefficientsDicCache;
+
+        public MockCoefficientProvider()
+        {
+            _coefficientsDicCache = new Dictionary<int, IList<int[]>>();
```
Comment (loning) [inline test/AElf.Kernel.SmartContract.ExecutionPluginForMethodFee.Tests/MockCoefficientProvider.cs]: How many times you copy the same code???

File: test/AElf.Contracts.TestContract.Tests/MockCoefficientProvider.cs [hunk 13]
```diff
@@ -0,0 +1,69 @@
...
+
+namespace AElf.Contract.TestContract
+{
+    public class MockCoefficientProvider : ICoefficientsCacheProvider, ISingletonDependency
+    {
+        private readonly Dictionary<int, IList<int[]>> _coefficientsDicCache;
+
+        public MockCoefficientProvider()
+        {
+            _coefficientsDicCache = new Dictionary<int, IList<int[]>>();
```
Comment (loning) [inline test/AElf.Contracts.TestContract.Tests/MockCoefficientProvider.cs]: Another copy

File: src/AElf.Kernel.FeeCalculation/AElf.Kernel.FeeCalculation.csproj [hunk 14]
```diff
@@ -0,0 +1,23 @@
...
+        <PackageId>AElf.Kernel.FeeCalculation</PackageId>
+        <GeneratePackageOnBuild>true</GeneratePackageOnBuild>
+        <Description>Fee Calculation Project.</Description>
+    </PropertyGroup>
+
+    <ItemGroup>
+      <ProjectReference Include="..\AElf.Kernel.SmartContract\AElf.Kernel.SmartContract.csproj" />
+      <ProjectReference Include="..\AElf.Kernel.Token\AElf.Kernel.Token.csproj" />
+      <!-- Have to ref SDK because this project need EventExtensions.ToLogEvent -->
+      <ProjectReference Include="..\AElf.Sdk.CSharp\AElf.Sdk.CSharp.csproj" />
```
Comment (loning) [inline src/AElf.Kernel.FeeCalculation/AElf.Kernel.FeeCalculation.csproj]: move  EventExtensions.ToLogEvent to AElf.CSharp.Core
reference AElf.CSharp.Core

**Pull request merged**